src/io.c

```diff
@@ -1432,9 +1432,26 @@ _dispatch_fd_entry_create_with_fd(dispatch_fd_t fd, uintptr_t hash)
 #if defined(_WIN32)
 		DWORD dwType = GetFileType((HANDLE)fd);
 		if (dwType == FILE_TYPE_PIPE) {
-			unsigned long value = 1;
-			int result = ioctlsocket((SOCKET)fd, (long)FIONBIO, &value);
-			(void)dispatch_assume_zero(result);
+			if (_dispatch_handle_is_socket((HANDLE)fd)) {
+				unsigned long value = 1;
+				int result = ioctlsocket((SOCKET)fd, (long)FIONBIO, &value);
+				(void)dispatch_assume_zero(result);
+			}
+			else {
+				// Try to make writing nonblocking, although pipes not coming
+				// from Foundation.Pipe may not have FILE_WRITE_ATTRIBUTES.
+				DWORD pipe_mode = PIPE_NOWAIT;
+				if (!SetNamedPipeHandleState((HANDLE)fd, &pipe_mode, NULL,
```

Comment:
Would this potentially clear the `PIPE_READMODE_MESSAGE`?  If so, should we do a `GetNamedPipeHandleState` first?

Reply:
Good point, that seems safer